plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Utils.hs

```diff
@@ -414,12 +416,15 @@ shouldUnconditionallyInline s n rhs body = preUnconditional ||^ postUnconditiona
     -- We actually also inline 'Dead' binders (i.e., remove dead code) here.
     preUnconditional = do
       isTermPure <- checkPurity rhs
-      nameUsedAtMostOnce n &&^ effectSafe body s n isTermPure
+      nameUsedAtMostOnce n &&^ (pure safe ||^ effectSafe body s n isTermPure)
```

Comment:
Hm, I don't think this is generally an improvement? `effectSafe` doesn't just check for effects, it also (somewhat implicitly) checks that the body "isn't doing any substantial work":

```haskell
    -- While we don't check for lambda etc like in the paper, `effectSafe` ensures that it
    -- isn't doing any substantial work.
```

With this change you may end up inlining something that does do substantial work. Although I guess not with this PR specifically. But it still may end up being disadvantageous.